### DISCUSSION TITLE
Dependency Injection on ViewModels

### DISCUSSION THREAD

**Discussion by jaymarvels** [AvaloniaUI/Avalonia#15630] [Q&A] [answered]
I have a User control (ReacticeUserControl<T>) that I created a ViewModel for:

```public class MyViewModel : ViewModelBase
{
    public ReactiveCommand<Unit, Unit> Exit { get; }

    public MyViewModel() 
    {
        Exit = ReactiveCommand.Create(ExitApplication);
    }

    private void ExitApplication() 
    {
        Locator.Current.GetService<IApplication>()?.Shutdown();
    }
}
```
I instantiate it in the ReacticeUserControl code behind:

```    protected override void OnLoaded(RoutedEventArgs e) 
    {
        //window = TopLevel.GetTopLevel(this) as Window;
        window = this.GetVisualRoot() as Window;
        this.DataContext = new MyViewModel();
        base.OnLoaded(e);
    }
```

All good so far.

If I now want to inject a dependency into the ViewModel like so:

```    private readonly IRepository<CityModel> _cityRepository;

    public MyViewModel(IRepository<CityModel> cityRepository) 
  {
        _cityRepository = cityRepository;
        Exit = ReactiveCommand.Create(ExitApplication);
    }
```

This means when I try to instantiate it requires this parameter and kind of defeats the point of DI. In my App code behind I have set up the configuration needed:

```
    public  App() 
    {
        // Let's do this in the constructor
        AppHost = Host.CreateDefaultBuilder()
            .UseContentRoot(AppContext.BaseDirectory)
            .ConfigureServices((context, services) => {
                // Register services with the DI container here.
                services.ConfigureFrameworkServices();
            })
            .Build();
    }
```

Just feel like I am missing or have missed a step somewhere, trying to keep things lightweight and only use ViewModels where needed

**Comment by timunie** [comment]
check Avalonia.Samples -> ViewInteractionSample

**Reply by jaymarvels** [reply]
Appreciate the swift response but in that example none of the viewmodels have any dependencies so unsure how this relates to my issue

**Reply by timunie** [reply]
Thought you wanted to exit your App. Otherwise DI works as it does anywhere, for example using MS DI or RxUI Splat

**Reply by jaymarvels** [reply]
Hi Tim,

Sorry I must not have explained well.

Exit app is all good, my issues are around constructor injection.

I register my services: `services.ConfigureFrameworkServices();` , which happens to include `services.AddTransient<IRepository<CityModel>, CityRepository>();`

In my ViewModel I can inject:

```private readonly IRepository<CityModel> _cityRepository;

    public MyViewModel(IRepository<CityModel> cityRepository) 
{
        _cityRepository = cityRepository;
        Exit = ReactiveCommand.Create(ExitApplication);
    }
```

Problem is when trying to instantiate the ViewModel in the View code behind: `this.DataContext = new MyViewModel();` with the injected constructor here at the code behind point its expecting "CityRepository" which defeats the point of using dependency injection.

Feels like I am almost there but just missing something or a step

**Reply by jaymarvels** [reply, marked as answer]
I think this is my issue, coming from a asp.net core background I was just expecting after some initial set up against the `Host.CreateDefaultBuilder()` that that would be enough, but looking through the docs and watching AngelSix it looks like the DI service registration is needed within the `OnFrameworkInitializationCompleted`

**Reply by thevortexcloud** [reply]
You can technically register the services elsewhere (EG in the `AfterSetup` method). But that is the easiest place to do it.

**Comment by LSXAxeller** [comment]
you can follow the avalonia docs guide []( https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection) , I applied it in Program.cs like this, and it's working like charm
```
public static void Main(string[] args)
{
    Log.Logger = new LoggerConfiguration()
        .Enrich.FromLogContext()
        .WriteTo.File("EGOIST_Handler.log", rollingInterval: RollingInterval.Month)
        .CreateLogger();

    var services = new ServiceCollection();
    var provider = services
        .AddLogging(loggingBuilder => loggingBuilder.AddSerilog(dispose: true))
        .AddKeyedSingleton<ITextService, ChatService>("ChatService")
        .AddKeyedSingleton<IModelsRepository, LocalModelsRepository>("LocalModelsRepository")
        .AddKeyedSingleton<IModelsRepository, HuggingFaceModelsRepository>("HuggingFaceModelsRepository")
        .AddKeyedSingleton<IModelsRepository, CivitAIModelsRepository>("CivitAIModelsRepository")
        .AddSingleton<MainWindowViewModel>()
        .AddSingleton<MainWindow>()
        .AddTransient<HomePageViewModel>()
        .AddTransient<HomePageView>()
        .AddTransient<TextPageViewModel>()
        .AddTransient<ChatPageViewModel>()
        .AddTransient<ChatPageView>()
        .BuildServiceProvider();
    Ioc.Default.ConfigureServices(provider);

    BuildAvaloniaApp()
    .StartWithClassicDesktopLifetime(args);
}
```

and if needed, you can get your services from anywhere in code like `Ioc.Default.GetService<MainWindowViewModel>();`

**Reply by thevortexcloud** [reply]
You are doing some things the Avalonia guide does not. Specifically you are setting up a static locator. [The original guide](https://github.com/AvaloniaUI/avalonia-docs/pull/359) _did_ do this as when I wrote the code I designed it to work across projects so it would work on every platform Avalonia supports. But what's there now only shows you how to set up DI without a locator (which is how you should be using it anyway in 99% of cases) in a single project. The locator should only ever be used when you have no other option as it creates lots of coupling.

**Reply by LSXAxeller** [reply]
I don't use the locator myself, as I wrote if needed, I am prefering to inject like I do in ASP.NET, in addition that I write in Program.cs  since I am targeting Desktop only so I have one presentation project, but thanks for reminding me about the coupling it cause.

**Reply by duongphuhiep** [reply]
After many research, my conclusion is that [Avalonia doesn't support Dependency Injection](https://github.com/AvaloniaUI/Avalonia/discussions/16775).

This is why the "unfamous" Service Locator is unavoidable in every Avalonia Application. 

@LSXAxeller: your `Ioc.Default` IS the unfamous Service Locator!  If I'm not wrong, you have to use this locator everywhere to "locate" your services:

```C#
HomePageView.DataContext = Ioc.Default.GetService<HomePageViewModel>();
TextPageView.DataContext = Ioc.Default.GetService<TextPageViewModel>();
ChatPageView.DataContext = Ioc.Default.GetService<ChatPageViewModel>();
```

It is not really "dependency inject", but as I said, we have no other choice. Unlike ASP.NET, [Avalonia doesn't support DI]((https://github.com/AvaloniaUI/Avalonia/discussions/16775)).

**Reply by thevortexcloud** [reply]
Avalonia does not have native support for DI. It is however not hard to add it yourself. You do not need to use the locator in most instances if you are only injecting view models (in fact I have written several programs in Avalonia that use DI without using a locator at all). Just follow the guide to create your root DI object.

https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection

Your root view model can have dependencies on other view models which you can pass around to where they are needed.

There are _some_ places you need to use a locator pattern though. EG in markup extensions as there is no way to pass a reference to the DI container otherwise.
